### PR TITLE
ARROW-12111: [Java] Generate flatbuffer files using flatc 1.12.0

### PR DESCRIPTION
**Pull request by nbauernfeind**
This is a resolution of ARROW-12111, which came up on the dev mailing list titled "[Java] Source control of generated flatbuffers code", the main complaint was that the plugin used to generate the flatbuffer files was unavailable for windows developers.

I, personally, would also like to see this upgraded to flatbuffers 1.12 from 1.9.

### CODE REVIEW

**Comment by kiszk** [comment]
Let me jump in this PR. flatbuffers 1.9 for s390x is not officially prepared. Thus, we privately build and provide it on own artifact. You can see [here](https://github.com/apache/arrow/blob/master/ci/scripts/java_build.sh#L33-L40).

If this approach works well, can we remove this workaround for s390x, too?

**Comment by nbauernfeind** [comment]
@kiszk I believe so! With this patch there is no need to have access to a `flatc` binary during any phase of the Java build. Shall I axe-those lines in the `java_build.sh` file you've pointed at?

File: java/pom.xml [hunk 1]
```diff
@@ -36,8 +36,8 @@
     <dep.netty.version>4.1.48.Final</dep.netty.version>
     <dep.jackson.version>2.11.4</dep.jackson.version>
     <dep.hadoop.version>2.7.1</dep.hadoop.version>
-    <dep.fbs.version>1.9.0</dep.fbs.version>
-    <dep.flatc.version>1.9.0</dep.flatc.version>
+    <dep.fbs.version>1.12.0</dep.fbs.version>
+    <dep.flatc.version>1.12.0</dep.flatc.version>
```
Comment (emkornfield) [inline java/pom.xml:40]: is this still used?

File: java/README.md [hunk 2]
```diff
@@ -64,26 +64,46 @@ and arrow-format into a single JAR.  Using the classifier "shade-format-flatbuff
...
+
+# remove the existing files
+rm -rf java/format/src
+
+# regenerate from the .fbs files
+flatc --java -o java/format/src/main/java format/*.fbs
+
+# prepend license header
+find java/format/src -type f | while read file; do
+  (cat header | while read line; do echo "// $line"; done; cat $file) > $file.tmp
```
Comment (emkornfield) [inline java/README.md:82]: is header checked in?  maybe add a line on how it was extracted?

File: java/README.md [hunk 3]
```diff
@@ -64,26 +64,46 @@ and arrow-format into a single JAR.  Using the classifier "shade-format-flatbuff
...
+To update the checked-in flatbuffer generated code perform the following:
+
+```bash
+cd $ARROW_HOME
+
+# remove the existing files
+rm -rf java/format/src
+
+# regenerate from the .fbs files
+flatc --java -o java/format/src/main/java format/*.fbs
```
Comment (emkornfield) [inline java/README.md:78]: it is probably good to document version here?

**Comment by emkornfield** [comment]
LGTM once CI passes.

File: java/README.md (hunk 2 above)
Comment (nbauernfeind) [reply java/README.md:82]: The header is already checked into the root of apache/arrow.

File: java/pom.xml (hunk 1 above)
Comment (nbauernfeind) [reply java/pom.xml]: The `dep.flatc.version` is not; good catch. (the `dep.fbs.version` is, however). I have force pushed the change to remove it.

File: java/README.md (hunk 3 above)
Comment (nbauernfeind) [reply java/README.md:88]: updated the readme to include a step to verify `flatc` matches the declared dependency version

**Comment by kiszk** [comment]
@nbauernfeind Let us wait for dropping it before seeing the test result.   

I am curious whether the Java code generated on the little-endian platform also works for the big-endian platform.

**Comment by bobtins** [comment]
@kiszk if the generated code didn't work on big-endian then it would be a flatbuffers bug because it's in charge of enforcing the binary format.
@nbauernfeind thanks for jumping in! Sorry for not following through on this; I got pulled off on other things.

**Review by bobtins — APPROVED** [review]
I've looked over this and it looks fine. I did spend some time on making a shell script that automatically does the manual steps you documented; I will give myself another 30 minutes to get that working, in which case I can just attach it somewhere and you can incorporate it if you wish.  I am going ahead and approving it so feel free to continue; the script is nice to have but you've documented the steps well.

**Pull request closed without merging**